Implement dynamic class provenance tracking to fix isinstance semantics and add support for dynamically defined enums

ogrisel opened the pull request:
This is a fix for #244 (and #101) to add support for dynamically defined Enum subclasses.

Properly adding support for dynamic Enums required to more broadly fix the `isinstance` semantics as initially requested in #195.

The proposed solution involves tracking the provenance of pickled dynamic class definitions with a pair of `weakref.WeakKeyDictionary` / `weakref.WeakValueDictionary` protected by a `threading.Lock`.

ogrisel commented:
@pitrou @pierreglaser if you have any suggestions / comments on this solution, please let me know.

ogrisel commented on tests/cloudpickle_test.py:
```diff
@@ -1257,6 +1258,44 @@ def test_dataclass(self):
+        # cloudpickle systematically creates new copies for locally defined
+        # classes that cannot be imported by name:
+        assert green1 is not Color.GREEN
+        assert ClonedColor is not Color
```
While the above is technically expected and respects the usual cloudpickle semantics, I believe this will be even more surprising than just "breaking isinstance" (as reported in #195): testing physical identity is actually the expected way to use enums.

In particular, this would mean that the following code that uses interactively defined Enums on a cluster would not work:

```python
from enum import Enum
from dask.distributed import Client

class Color(Enum):
    RED = 1
    GREEN = 2

def check_data(data):
    return Color.GREEN if sum(data) > 0 else Color.RED

client = Client()
client.submit(check_data, [1, 2, 3]).result() is Color.GREEN
```

Therefore I am not sure if supporting interactively defined enums is a good feature for cloudpickle or not...

pitrou replied:
Shouldn't "interactively defined enums" be fetched from `__main__` rather than recreated?

ogrisel replied:
Maybe and this should probably be the case for any interactively defined types in `__main__` (see #195) but we should be careful to still make sure that, if later, that implementation changes in the main process, no previous copies will lie around in the `__main__` of worker processes, otherwise the new implementation will never get propagated to the worker process:


```
from enum import Enum
from dask.distributed import Client

class Color(Enum):
    RED = 1
    GREEN = 2

def check_data(data):
    return Color.GREEN if sum(data) >= 0 else Color.RED

client = Client()
client.submit(check_data, [1, 2, 3]).result() is Color.GREEN

class Color(Enum):
    RED = 1
    BLUE = 2

def check_data(data):
    return Color.BLUE if sum(data) >= 0 else Color.RED

client.submit(check_data, [1, 2, 3]).result() is Color.BLUE
```

We currently lack a missing test harness to add non-regression test for those semantics in the cloudpickle test suite. The `subprocess_pickle_echo` thingy does not allow to check for the typical interactive remote call use-case of pyspark / loky / dask-distributed / ray justifies the existence of cloudpickle in the first place. I think improving the test harness in that respect is a priority.

ogrisel replied:
I run some interactive tests with loky and the current cloudpickle master, and nothing is leaked (`setattr`'ed on sys.modules["__main__"]) so this should be fine.

Note that locally defined enums as tested in the test of this PR would still be non-lookup-able but I think it would be fine to special case the `__main__` module.

ogrisel replied:
I pushed a much better solution that tracks any dynamic class definition and matched unpickled instances back to them.

ogrisel commented on tests/cloudpickle_test.py:
```diff
@@ -1257,6 +1258,66 @@ def test_dataclass(self):
+        assert green1 is green2
+        assert green1 is ClonedColor.GREEN
+
+        # Check that the pickle that was return has been matched back to the
```
```suggestion
        # Check that the pickle that was returned has been matched back to the
```

ogrisel commented:
I had forgotten about #105 which implemented similar support for `Enum` classes in `__main__` as first requested in #101.

However I also think this is a dangerous feature in its current implementation. I think we need to better think trough when an how to reconcile dynamic types (including enums for also other dynamically defined types) on a cluster with some explicit scoping mechanism.

ogrisel commented:
Here is another valid use case that cannot be support by this PR:

```python
from enum import Enum
from dask.distributed import Client


class Color(Enum):
    RED = 1
    GREEN = 2


client = Client()
data_futures = client.scatter([Color.RED, Color.GREEN])

def is_red(x):
    return x is Color.RED

client.map(is_red, data_futures)
```

`data_futures` and `is_red` are pickled each by their own `CloudPickler` instances with their own memoizer and global_ref namespaces. The reconstructed Color classes on the worker are therefore independent and `is_red` can never return `True` under those conditions.

ogrisel commented on cloudpickle/cloudpickle.py:
```diff
@@ -460,6 +460,31 @@ def func():
+            extra["qualname"] = obj.__qualname__
+
+        self.save_reduce(_make_dynamic_enum,
+                         (obj.__base__, obj.__name__, elements, doc,
```
It should be `obj.__bases__[1]` I think.  `obj.__bases__[-1]` is going to be `enum.EnumMeta` in this case and this is probably not what we want here.

pierreglaser replied:
> `obj.__bases__[-1]` is going to be `enum.EnumMeta`

`enum.EnumMeta` is the type of `obj`, no? I think `obj.__bases__` contains the class inherited by `obj` (like `enum.IntEnum` for instance).

> It should be `obj.__bases__[1]`

I think (athough not 100% sure) that `obj.__bases__` can only be of size 1 or 2 for `Enum` classes, to in that case that would be equivalent.

llllllllll replied:
I think we need to forward all of the bases. This list can be as long as you want, each type is a mixin added to the values of the enum. For example:

```python
In [4]: class A:
   ...:     def a(self): return 'a'
   ...:     

In [5]: class B:
   ...:     def b(self): return 'b'
   ...:     

In [6]: class ABEnum(A, B, enum.Enum):
   ...:     v1 = 1
   ...:     v2 = 2
   ...:     

In [7]: ABEnum.v2.a()
Out[7]: 'a'

In [8]: ABEnum.v2.b()
Out[8]: 'b'

In [9]: ABEnum.v2
Out[9]: <ABEnum.v2: 2>
```

The current implementation of `_make_dynamic_enum` will fail if it needs to serialize a class like `IntEnum` (note: not a subclass of `IntEnum`, `IntEnum` itself) without being able to use the `__module__` and `__name__` style class serialization. here is a failing test that demonstrates this:

```Python
    def test_locally_defined_string_enum(self):
        enum = pytest.importorskip("enum")

        class StringEnum(str, enum.Enum):
            """Enum when all members are also (and must be) strings"""

        class Color(StringEnum):
            RED = 'RED'
            GREEN = 'GREEN'
            BLUE = 'BLUE'

        green1, green2, ClonedColor = pickle_depickle(
            [Color.GREEN, Color.GREEN, Color],
            protocol=self.protocol)

        assert green1 is green2
        assert green1 is ClonedColor.GREEN
        assert green1 is not ClonedColor.BLUE
        assert ClonedColor is Color
```

`EnumMeta` does ensure that the `Enum` class part does come _last_ (https://github.com/python/cpython/blob/master/Lib/enum.py#L496-L499) so we at least know that can rely on that. I would recommend sending the entire class dict (pruning some unneeded things like non-overridden methods) and the entire `__bases__` to be reconstructed on the receiving end. Another reason to serialize the entire class body is that enums may have methods, for example a custom `__repr__`. Currently, we do not send extra methods:


```Python
    def test_interactively_defined_enum_with_method(self):
        pytest.importorskip("enum")
        code = """if __name__ == "__main__":
        from enum import Enum
        from testutils import subprocess_worker

        with subprocess_worker(protocol={protocol}) as w:

            class WithMethod(Enum):
                x = 2
                y = 3

                def method(self):
                    return self.value * 2

            def check(x1, x2, ClonedWithMethod):
                assert x1 is x2
                assert x1 is ClonedWithMethod.x
                assert x1 is not ClonedWithMethod.y

                assert x1.method() == 4

            w.run(check, WithMethod.x, WithMethod.x, WithMethod)
        """.format(protocol=self.protocol)
        assert_run_python_script(code)
```

pierreglaser replied:
> I think we need to forward all the bases

I agree, however this current implementation uses `EnumMeta._create_`, which only forwards two base classes: the class on which `_create_` was called, and the `type` argument. The reason why @ogrisel ended up doing it this way (instead of using `types.new_class`) is explained here, https://github.com/cloudpipe/cloudpickle/issues/244#issuecomment-462066774.

ogrisel commented:
@mrocklin you might be interested in this PR. Basically it makes it possible to properly support enums in dask and fix the isinstance semantics for object passed as futures.

@pitrou @mrocklin I would also love to get your feedback on the solution I used to tackle this issue that involves a WeakKeyDictionary / WeakValueDictionary pair protected by a lock. Maybe there is a better way to achieve a similar result.

@pcmoritz @HyukjinKwon I would love it if you could test this PR with Ray and Spark to check that it does not introduce any regression.

ogrisel commented:
@mrocklin when launching the dask-distributed tests against this PR,  `test_prometheus` fails:

https://travis-ci.org/cloudpipe/cloudpickle/jobs/494254377

It does not seem to be related to the change of this PR but I am not 100% sure and it seems deterministic.

mrocklin commented:
Resolved here: https://github.com/dask/distributed/pull/2533

ogrisel commented:
Alright: all dowstream CI tests pass (dask, loky, joblib). I removed the hard dependency to the enum module so that it should fix a broken test on Py 2.7 on Ray. Waiting for more feedback.

ogrisel commented:
@llllllllll @ssanderson I would also love to get your input on this PR.

pcmoritz commented:
From the Ray side, this PR passes the test suite, so looking forward to seeing it merged!

Thanks for all the hard work :)

ogrisel commented on cloudpickle/cloudpickle.py:
```diff
@@ -1143,6 +1208,22 @@ def _make_skel_func(code, cell_count, base_globals=None):
+    definition under that id so that other instances stemming from the same
+    class id will also reuse this class definition.
+
+    The extra variable is meant to be a dict use for forward compatibility
```
Good idea.

ogrisel commented on cloudpickle/cloudpickle.py:
```diff
@@ -1143,6 +1208,22 @@ def _make_skel_func(code, cell_count, base_globals=None):
+                         class_tracker_id, extra):
+    """Build dynamic class with an empty __dict__ to be filled once memoized
+
+    If class_tracker_id is not None, try to lookup an existing class definition
```
It used to be the case in a former version of this PR. However I wonder if it's not a good idea to make it possible to disable tracking by passing None for forward compatibility.

I am not sure. Would this be a YAGNI?

llllllllll replied:
Hmm, this seems fine given that it is already implemented.

ogrisel commented on cloudpickle/cloudpickle.py:
```diff
@@ -468,6 +527,9 @@ def save_dynamic_class(self, obj):
         from global modules.
         """
+        if Enum is not None and issubclass(obj, Enum):
+            return self._save_dynamic_enum(obj)
```
Maybe that's what the `type` argument is for. I forgot to address @pierreglaser's comments above.

ogrisel replied:
I had a deeper look with @pierreglaser and it seems that this is a fundamental limitation of enums that inherit from another Python class: this problem is already present for enums defined statically in importable modules. I therefore believe this cannot be fixed in cloudpickle but should be fixed upstream in Python's pickle or enum modules themselves.

ogrisel replied:
To be more specific:

```python

In [6]: cat mymodule.py
from enum import Enum


class A:
    def method_a(self):
        return "a"


class E(A, Enum):
    x = 1
    y = 2



In [7]: from mymodule import E

In [8]: import pickle

In [9]: pickle.dumps(E.x)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-5c82fc8d827e> in <module>()
----> 1 pickle.dumps(E.x)

/opt/python3.7/lib/python3.7/enum.py in _break_on_call_reduce(self, proto)
     42     """Make the given class un-picklable."""
     43     def _break_on_call_reduce(self, proto):
---> 44         raise TypeError('%r cannot be pickled' % self)
     45     cls.__reduce_ex__ = _break_on_call_reduce
     46     cls.__module__ = '<unknown>'

TypeError: <E.x: 1> cannot be pickled
```

The class `A` can be imported from `mymodule` so maybe it's actually a bug in Python itself.

pierreglaser replied:
I think somehow this is not the case anymore in recent python versions (3.7.1 and later):
```python
Python 3.7.1 (tags/v3.7.1:260ec2c36a, Jan 24 2019, 18:18:14)                                                                                                                                                       
Type 'copyright', 'credits' or 'license' for more information                                                                                                                                                      
IPython 7.5.0 -- An enhanced Interactive Python. Type '?' for help.                                                                                                                                                
                                                                                                                                                                                                                   
In [1]: class A:                                                                                                                                                                                                   
   ...:     def a(self): return "a"                                                                                                                                                                                
   ...: import enum                                                                                                                                                                                                
   ...: class AEnum(A, enum.Enum):                                                                                                                                                                                 
   ...:     value = 1                                                                                                                                                                                              
   ...: import pickle                                                                                                                                                                                              
   ...: pickle.dumps(AEnum.value)                                                                                                                                                                                  
Out[1]: b'\x80\x03c__main__\nAEnum\nq\x00K\x01\x85q\x01Rq\x02.'     
```

ogrisel commented on cloudpickle/cloudpickle.py:
```diff
@@ -460,6 +492,33 @@ def func():
+
+        extra = {}
+        if hasattr(obj, "__qualname__"):
+            extra["qualname"] = obj.__qualname__
```
The problem is that Python 2.7 does not have `__qualname__` but can have enums via the `enum34` backport.

ogrisel commented:
It seems that my latest commits have broken some dask tests although I am not 100% sure because there is no output for 10min and travis kills the job without letting pytest report the failures. Will have to do that manually but I won't have time soon enough. If someone else wants to have a look don't hesitate.

ogrisel commented:
@llllllllll I have rebased this PR to check that the downstream tests of dask distributed, ray, loky and joblib pass.

I now pass all the bases as you suggested so and also added support for methods defined directly on the enum class itself.

pierreglaser commented on cloudpickle/cloudpickle.py:
```diff
@@ -1124,6 +1244,40 @@ def _rehydrate_skeleton_class(skeleton_class, class_dict):
+    # the list of base classes:
+    enum_base = bases[-1]
+    metacls = enum_base.__class__
+    _, first_enum = enum_base._get_mixins_(bases)
```
This line can be removed no?

ogrisel replied:
Good catch, fixed.

philippotto commented:
Sorry for nagging here, but is there anything which blocks this PR? The fix would be pretty useful for my team.

ogrisel commented:
I think we might want to release cloudpickle 1.0 without this change and 1.1 with this PR merged so that we have a pretty uncontroversial 1.0 release for cloudpickle as @mrocklin recently asked.

ogrisel commented:
@llllllllll I would like to merge this but I would really appreciate it if you could confirm that I took your comments into account as intended.

ogrisel commented:
Thanks you @llllllllll for your review. I am merging.

ogrisel commented:
@philippotto cloudpickle 1.1.0 is out with this fix included.

philippotto commented:
Awesome, thank you!
